drivers/phoenixcontact_modbus.c

```diff
@@ -21,727 +21,864 @@
+ 
+ #define QUINT4_10A_UPS_PARTNUMBER 2907067
+ #define QUINT4_10A_UPS_DESCRIPTION "QUINT4-UPS/24DC/24DC/10/USB"
+ 
+ #define QUINT4_20A_UPS_PARTNUMBER 2907072
+ #define QUINT4_20A_UPS_DESCRIPTION "QUINT4-UPS/24DC/24DC/20/USB"
+ 
+ #define QUINT4_40A_UPS_PARTNUMBER 2907078
+ #define QUINT4_40A_UPS_DESCRIPTION "QUINT4-UPS/24DC/24DC/40/USB"
+ 
+ #define TRIO_5A_UPS_PARTNUMBER 2866611
+ #define TRIO_5A_UPS_DESCRIPTION "TRIO-UPS/1AC/24DC/5"
+ 
+ #define TRIO_2G_5A_UPS_PARTNUMBER 2907160
+ #define TRIO_2G_5A_UPS_DESCRIPTION "TRIO-UPS-2G/1AC/24DC/5"
+ 
+ #define TRIO_2G_10A_UPS_PARTNUMBER 2907161
+ #define TRIO_2G_10A_UPS_DESCRIPTION "TRIO-UPS-2G/1AC/24DC/10"
+ 
+ #define TRIO_2G_20A_UPS_PARTNUMBER 1105556
+ #define TRIO_2G_20A_UPS_DESCRIPTION "TRIO-UPS-2G/1AC/24DC/20"
+ 
+ 
+ /* --- Shutdown & Mode Configuration --- */
+ 
+ // Macro for fetching a uint16_t config value with fallback
+ #define GETVAL_U16(name, fallback) (getval(name) ? (uint16_t)atoi(getval(name)) : (fallback))
+ 
+ // Time [s] after entering battery mode before shutdown signal (bit 15) is triggered
+ #define REG_PC_SHUTDOWN_DELAY        0x105A  // default: 60s, range: 1–65535
+ 
+ // Time [s] allowed for PC to shut down before output turns off
+ #define REG_PC_SHUTDOWN_TIME         0x105D  // default: 120s, range: 1–3600
+ 
+ // Time [s] output is off after PC shutdown before reboot
+ #define REG_PC_RESET_TIME            0x105E  // default: 10s, range: 0–60
+ 
+ // Time [s] input voltage must be above threshold to return to mains
+ #define REG_MAINS_RETURN_DELAY       0x1058  // default: 10s
+ 
+ // Selector mode switch: PC-Mode = 9
+ #define REG_MODE_SELECTOR_SWITCH     0x1074  // default: 0–9
+ 
+ /* --- Battery Monitoring --- */
+ 
+ // SOH warning threshold [%]
+ #define REG_WARNING_SOH_THRESHOLD    0x1071  // default: 0 (disabled), range: 1–100
+ 
+ // Switch to battery mode if below this input voltage [mV]
+ #define REG_VOLTAGE_BELOW_BATTERY    0x1056  // example: 21000
+ 
+ // Switch to mains mode if above this voltage [mV]
+ #define REG_VOLTAGE_ABOVE_MAINS      0x1057  // example: 23000
+ 
+ typedef enum
+ {
+	 NONE,
+	 QUINT_UPS,
+	 QUINT4_UPS,
+	 TRIO_UPS,
+	 TRIO_2G_UPS
+ } models;
+ 
+ /* Variables */
+ static modbus_t *modbus_ctx = NULL;
+ static int errcount = 0;
+ 
+ static int mrir(modbus_t * arg_ctx, int addr, int nb, uint16_t * dest);
+ 
+ static models UPSModel = NONE;
+ 
+ typedef struct {
+	 const char *nut_name;
+	 uint16_t reg_addr;
+ } delay_param_t;
+ 
+ static const delay_param_t delay_params[] = {
+	 { "ups.delay.shutdown",     REG_PC_SHUTDOWN_DELAY},
+	 { "ups.delay.pc_shutdown",  REG_PC_SHUTDOWN_TIME},
+	 { "ups.delay.pc_reset",     REG_PC_RESET_TIME},
+	 { "ups.delay.mains_return", REG_MAINS_RETURN_DELAY},
+	 { "ups.mode.selector",      REG_MODE_SELECTOR_SWITCH},
+	 { "ups.bat.warning_soh",    REG_WARNING_SOH_THRESHOLD},
+	 { "ups.bat.voltage_low",    REG_VOLTAGE_BELOW_BATTERY},
+	 { "ups.mains.voltage_high", REG_VOLTAGE_ABOVE_MAINS } 
```

Comment:
I believe these strings are also driver configuration parameters (per `addvar()` below)?

If any of these concepts are already described by standard NUT variable/command names listed in docs/nut-names.txt, it would be beneficial to add corresponding code so `upsrw`/`upscmd` can be used to set them at run-time, same way as for any other supported UPS with such features. But this may be improved later.